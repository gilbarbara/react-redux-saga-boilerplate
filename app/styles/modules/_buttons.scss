.btn-icon {
  align-items: center;
  display: inline-flex;
  line-height: 1;

  span {
    font-weight: 700;
    margin-left: $grid-gutter-width / 6;
    text-transform: uppercase;
  }
}
