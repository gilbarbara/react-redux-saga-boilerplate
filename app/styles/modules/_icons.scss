@font-face {
  font-family: 'icon-font-v2';
  font-style: normal;
  font-weight: normal;
  src: url('#{$font-path}icon-font-v2.ttf?hiqvyh') format('truetype'),
    url('#{$font-path}icon-font-v2.woff?hiqvyh') format('woff'),
    url('#{$font-path}icon-font-v2.svg?hiqvyh#icon-font-v2') format('svg');
}

[class^='i-'],
[class*=' i-'] {
  /* use !important to prevent issues with browser extensions that change fonts */
  display: inline-block;
  font-family: 'icon-font-v2' !important;
  -moz-osx-font-smoothing: grayscale;
  -webkit-font-smoothing: antialiased;
  font-style: normal;
  font-variant: normal;
  font-weight: normal;
  line-height: 1;
  speak: none;
  text-transform: none;
}

// *** Animations *** //
.i-spin {
  animation: rotate 2s infinite linear;
}

.i-fade {
  animation: fadeLoop 1.2s infinite linear;
}

.i-pulse {
  animation: pulse 1.2s infinite linear;
}

// *** Icons *** //
.i-album:before {
  content: '\f000';
}

.i-albums:before {
  content: '\f001';
}

.i-anchor:before {
  content: '\f002';
}

.i-angle-double-down:before {
  content: '\f003';
}

.i-angle-double-left:before {
  content: '\f004';
}

.i-angle-double-right:before {
  content: '\f005';
}

.i-angle-double-up:before {
  content: '\f006';
}

.i-angle-down:before {
  content: '\f007';
}

.i-angle-left:before {
  content: '\f008';
}

.i-angle-right:before {
  content: '\f009';
}

.i-angle-up:before {
  content: '\f00a';
}

.i-archive:before {
  content: '\f00b';
}

.i-area-chart:before {
  content: '\f00c';
}

.i-arrow-circle-down:before {
  content: '\f00d';
}

.i-arrow-circle-left:before {
  content: '\f00e';
}

.i-arrow-circle-o-down:before {
  content: '\f00f';
}

.i-arrow-circle-o-left:before {
  content: '\f010';
}

.i-arrow-circle-o-right:before {
  content: '\f011';
}

.i-arrow-circle-o-up:before {
  content: '\f012';
}

.i-arrow-circle-right:before {
  content: '\f013';
}

.i-arrow-circle-up:before {
  content: '\f014';
}

.i-arrow-down:before {
  content: '\f015';
}

.i-arrow-left:before {
  content: '\f016';
}

.i-arrow-right:before {
  content: '\f017';
}

.i-arrow-up:before {
  content: '\f018';
}

.i-arrows:before {
  content: '\f019';
}

.i-arrows-alt:before {
  content: '\f01a';
}

.i-arrows-h:before {
  content: '\f01b';
}

.i-arrows-v:before {
  content: '\f01c';
}

.i-artists:before {
  content: '\f01d';
}

.i-asterisk:before {
  content: '\f01e';
}

.i-at:before {
  content: '\f01f';
}

.i-automobile:before {
  content: '\f020';
}

.i-car:before {
  content: '\f020';
}

.i-backward:before {
  content: '\f021';
}

.i-ban:before {
  content: '\f022';
}

.i-bar-chart:before {
  content: '\f023';
}

.i-bar-chart-o:before {
  content: '\f023';
}

.i-bars:before {
  content: '\f024';
}

.i-navicon:before {
  content: '\f024';
}

.i-reorder:before {
  content: '\f024';
}

.i-bell:before {
  content: '\f025';
}

.i-bell-o:before {
  content: '\f026';
}

.i-bell-slash:before {
  content: '\f027';
}

.i-bell-slash-o:before {
  content: '\f028';
}

.i-bicycle:before {
  content: '\f029';
}

.i-binoculars:before {
  content: '\f02a';
}

.i-bolt:before {
  content: '\f02b';
}

.i-flash:before {
  content: '\f02b';
}

.i-bomb:before {
  content: '\f02c';
}

.i-book:before {
  content: '\f02d';
}

.i-bookmark:before {
  content: '\f02e';
}

.i-bookmark-o:before {
  content: '\f02f';
}

.i-bug:before {
  content: '\f030';
}

.i-building:before {
  content: '\f031';
}

.i-bullhorn:before {
  content: '\f032';
}

.i-bullseye:before {
  content: '\f033';
}

.i-bus:before {
  content: '\f034';
}

.i-cab:before {
  content: '\f035';
}

.i-taxi:before {
  content: '\f035';
}

.i-calendar:before {
  content: '\f036';
}

.i-calendar-check-o:before {
  content: '\f037';
}

.i-calendar-minus-o:before {
  content: '\f038';
}

.i-calendar-o:before {
  content: '\f039';
}

.i-calendar-plus-o:before {
  content: '\f03a';
}

.i-calendar-times-o:before {
  content: '\f03b';
}

.i-camera:before {
  content: '\f03c';
}

.i-camera-retro:before {
  content: '\f03d';
}

.i-caret-down:before {
  content: '\f03e';
}

.i-caret-left:before {
  content: '\f03f';
}

.i-caret-right:before {
  content: '\f040';
}

.i-caret-up:before {
  content: '\f041';
}

.i-cart-arrow-down:before {
  content: '\f042';
}

.i-cart-plus:before {
  content: '\f043';
}

.i-certificate:before {
  content: '\f044';
}

.i-chain:before {
  content: '\f045';
}

.i-link:before {
  content: '\f045';
}

.i-chain-broken:before {
  content: '\f046';
}

.i-unlink:before {
  content: '\f046';
}

.i-check:before {
  content: '\f047';
}

.i-check-circle:before {
  content: '\f048';
}

.i-check-circle-o:before {
  content: '\f049';
}

.i-check-square:before {
  content: '\f04a';
}

.i-check-square-o:before {
  content: '\f04b';
}

.i-chevron-circle-down:before {
  content: '\f04c';
}

.i-chevron-circle-left:before {
  content: '\f04d';
}

.i-chevron-circle-right:before {
  content: '\f04e';
}

.i-chevron-circle-up:before {
  content: '\f04f';
}

.i-chevron-down:before {
  content: '\f050';
}

.i-chevron-left:before {
  content: '\f051';
}

.i-chevron-right:before {
  content: '\f052';
}

.i-chevron-up:before {
  content: '\f053';
}

.i-child:before {
  content: '\f054';
}

.i-circle-o:before {
  content: '\f055';
}

.i-circle-o-notch:before {
  content: '\f056';
}

.i-clock-o:before {
  content: '\f057';
}

.i-close:before {
  content: '\f058';
}

.i-remove:before {
  content: '\f058';
}

.i-times:before {
  content: '\f058';
}

.i-cloud-download:before {
  content: '\f059';
}

.i-cloud-upload:before {
  content: '\f05a';
}

.i-code:before {
  content: '\f05b';
}

.i-code-fork:before {
  content: '\f05c';
}

.i-coffee:before {
  content: '\f05d';
}

.i-cog:before {
  content: '\f05e';
}

.i-gear:before {
  content: '\f05e';
}

.i-cogs:before {
  content: '\f05f';
}

.i-gears:before {
  content: '\f05f';
}

.i-columns:before {
  content: '\f060';
}

.i-comment:before {
  content: '\f061';
}

.i-comment-o:before {
  content: '\f062';
}

.i-commenting:before {
  content: '\f063';
}

.i-commenting-o:before {
  content: '\f064';
}

.i-comments:before {
  content: '\f065';
}

.i-copyright:before {
  content: '\f066';
}

.i-creative-commons:before {
  content: '\f067';
}

.i-credit-card:before {
  content: '\f068';
}

.i-credit-card-alt:before {
  content: '\f069';
}

.i-crosshairs:before {
  content: '\f06a';
}

.i-cutlery:before {
  content: '\f06b';
}

.i-dashboard:before {
  content: '\f06c';
}

.i-tachometer:before {
  content: '\f06c';
}

.i-desktop:before {
  content: '\f06d';
}

.i-dot-circle-o:before {
  content: '\f06e';
}

.i-download:before {
  content: '\f06f';
}

.i-edit:before {
  content: '\f070';
}

.i-pencil-square-o:before {
  content: '\f070';
}

.i-eject:before {
  content: '\f071';
}

.i-ellipsis-h:before {
  content: '\f072';
}

.i-ellipsis-v:before {
  content: '\f073';
}

.i-envelope:before {
  content: '\f074';
}

.i-envelope-o:before {
  content: '\f075';
}

.i-eraser:before {
  content: '\f076';
}

.i-exchange:before {
  content: '\f077';
}

.i-exclamation-circle:before {
  content: '\f078';
}

.i-exclamation-triangle:before {
  content: '\f079';
}

.i-warning:before {
  content: '\f079';
}

.i-expeditedssl:before {
  content: '\f07a';
}

.i-external-link:before {
  content: '\f07b';
}

.i-external-link-square:before {
  content: '\f07c';
}

.i-eye:before {
  content: '\f07d';
}

.i-eye-slash:before {
  content: '\f07e';
}

.i-eyedropper:before {
  content: '\f07f';
}

.i-facebook:before {
  content: '\f080';
}

.i-facebook-f:before {
  content: '\f080';
}

.i-facebook-square:before {
  content: '\f081';
}

.i-fast-backward:before {
  content: '\f082';
}

.i-fast-forward:before {
  content: '\f083';
}

.i-fax:before {
  content: '\f084';
}

.i-feed:before {
  content: '\f085';
}

.i-rss:before {
  content: '\f085';
}

.i-female:before {
  content: '\f086';
}

.i-file:before {
  content: '\f087';
}

.i-file-o:before {
  content: '\f088';
}

.i-film:before {
  content: '\f089';
}

.i-filter:before {
  content: '\f08a';
}

.i-fire:before {
  content: '\f08b';
}

.i-fire-extinguisher:before {
  content: '\f08c';
}

.i-flag:before {
  content: '\f08d';
}

.i-flag-checkered:before {
  content: '\f08e';
}

.i-flag-o:before {
  content: '\f08f';
}

.i-folder:before {
  content: '\f090';
}

.i-folder-o:before {
  content: '\f091';
}

.i-folder-open:before {
  content: '\f092';
}

.i-folder-open-o:before {
  content: '\f093';
}

.i-forward:before {
  content: '\f094';
}

.i-frown-o:before {
  content: '\f095';
}

.i-gavel:before {
  content: '\f096';
}

.i-legal:before {
  content: '\f096';
}

.i-gift:before {
  content: '\f097';
}

.i-github:before {
  content: '\f098';
}

.i-glass:before {
  content: '\f099';
}

.i-globe:before {
  content: '\f09a';
}

.i-google:before {
  content: '\f09b';
}

.i-google-plus:before {
  content: '\f09c';
}

.i-google-plus-circle:before {
  content: '\f09d';
}

.i-google-plus-official:before {
  content: '\f09d';
}

.i-group:before {
  content: '\f09e';
}

.i-users:before {
  content: '\f09e';
}

.i-hand-grab-o:before {
  content: '\f09f';
}

.i-hand-rock-o:before {
  content: '\f09f';
}

.i-hand-o-right:before {
  content: '\f0a0';
}

.i-hand-paper-o:before {
  content: '\f0a1';
}

.i-hand-stop-o:before {
  content: '\f0a1';
}

.i-hashtag:before {
  content: '\f0a2';
}

.i-hdd-o:before {
  content: '\f0a3';
}

.i-headphones:before {
  content: '\f0a4';
}

.i-heart:before {
  content: '\f0a5';
}

.i-heart-o:before {
  content: '\f0a6';
}

.i-history:before {
  content: '\f0a7';
}

.i-home:before {
  content: '\f0a8';
}

.i-hourglass:before {
  content: '\f0a9';
}

.i-hourglass-1:before {
  content: '\f0aa';
}

.i-hourglass-start:before {
  content: '\f0aa';
}

.i-hourglass-2:before {
  content: '\f0ab';
}

.i-hourglass-half:before {
  content: '\f0ab';
}

.i-hourglass-3:before {
  content: '\f0ac';
}

.i-hourglass-end:before {
  content: '\f0ac';
}

.i-hourglass-o:before {
  content: '\f0ad';
}

.i-image:before {
  content: '\f0ae';
}

.i-photo:before {
  content: '\f0ae';
}

.i-picture-o:before {
  content: '\f0ae';
}

.i-inbox:before {
  content: '\f0af';
}

.i-industry:before {
  content: '\f0b0';
}

.i-info-circle:before {
  content: '\f0b1';
}

.i-instagram:before {
  content: '\f0b2';
}

.i-key:before {
  content: '\f0b3';
}

.i-keyboard-o:before {
  content: '\f0b4';
}

.i-language:before {
  content: '\f0b5';
}

.i-laptop:before {
  content: '\f0b6';
}

.i-library:before {
  content: '\f0b7';
}

.i-lightbulb-o:before {
  content: '\f0b8';
}

.i-line-chart:before {
  content: '\f0b9';
}

.i-linkedin:before {
  content: '\f0ba';
}

.i-linkedin-square:before {
  content: '\f0bb';
}

.i-list-alt:before {
  content: '\f0bc';
}

.i-location-arrow:before {
  content: '\f0bd';
}

.i-lock:before {
  content: '\f0be';
}

.i-long-arrow-down:before {
  content: '\f0bf';
}

.i-long-arrow-left:before {
  content: '\f0c0';
}

.i-long-arrow-right:before {
  content: '\f0c1';
}

.i-long-arrow-up:before {
  content: '\f0c2';
}

.i-low-vision:before {
  content: '\f0c3';
}

.i-magic:before {
  content: '\f0c4';
}

.i-magnet:before {
  content: '\f0c5';
}

.i-mail-forward:before {
  content: '\f0c6';
}

.i-share:before {
  content: '\f0c6';
}

.i-mail-reply:before {
  content: '\f0c7';
}

.i-reply:before {
  content: '\f0c7';
}

.i-mail-reply-all:before {
  content: '\f0c8';
}

.i-reply-all:before {
  content: '\f0c8';
}

.i-male:before {
  content: '\f0c9';
}

.i-map:before {
  content: '\f0ca';
}

.i-map-marker:before {
  content: '\f0cb';
}

.i-map-o:before {
  content: '\f0cc';
}

.i-map-pin:before {
  content: '\f0cd';
}

.i-map-signs:before {
  content: '\f0ce';
}

.i-meh-o:before {
  content: '\f0cf';
}

.i-microphone:before {
  content: '\f0d0';
}

.i-microphone-slash:before {
  content: '\f0d1';
}

.i-minus:before {
  content: '\f0d2';
}

.i-minus-circle:before {
  content: '\f0d3';
}

.i-minus-square:before {
  content: '\f0d4';
}

.i-minus-square-o:before {
  content: '\f0d5';
}

.i-mixcloud:before {
  content: '\f0d6';
}

.i-mobile:before {
  content: '\f0d7';
}

.i-mobile-phone:before {
  content: '\f0d7';
}

.i-money:before {
  content: '\f0d8';
}

.i-moon-o:before {
  content: '\f0d9';
}

.i-music:before {
  content: '\f0da';
}

.i-offline:before {
  content: '\f0db';
}

.i-paint-brush:before {
  content: '\f0dc';
}

.i-paper-plane:before {
  content: '\f0dd';
}

.i-send:before {
  content: '\f0dd';
}

.i-pause:before {
  content: '\f0de';
}

.i-pause-circle:before {
  content: '\f0df';
}

.i-pause-circle-o:before {
  content: '\f0e0';
}

.i-paw:before {
  content: '\f0e1';
}

.i-pencil:before {
  content: '\f0e2';
}

.i-pencil-square:before {
  content: '\f0e3';
}

.i-percent:before {
  content: '\f0e4';
}

.i-phone:before {
  content: '\f0e5';
}

.i-phone-square:before {
  content: '\f0e6';
}

.i-pie-chart:before {
  content: '\f0e7';
}

.i-pinterest:before {
  content: '\f0e8';
}

.i-pinterest-p:before {
  content: '\f0e9';
}

.i-plane:before {
  content: '\f0ea';
}

.i-play:before {
  content: '\f0eb';
}

.i-play-circle:before {
  content: '\f0ec';
}

.i-play-circle-o:before {
  content: '\f0ed';
}

.i-playlist:before {
  content: '\f0ee';
}

.i-plug:before {
  content: '\f0ef';
}

.i-plus:before {
  content: '\f0f0';
}

.i-plus-circle:before {
  content: '\f0f1';
}

.i-plus-square:before {
  content: '\f0f2';
}

.i-power-off:before {
  content: '\f0f3';
}

.i-print:before {
  content: '\f0f4';
}

.i-puzzle-piece:before {
  content: '\f0f5';
}

.i-question-circle:before {
  content: '\f0f6';
}

.i-question-circle-o:before {
  content: '\f0f7';
}

.i-queue:before {
  content: '\f0f8';
}

.i-quote-left:before {
  content: '\f0f9';
}

.i-quote-right:before {
  content: '\f0fa';
}

.i-radio:before {
  content: '\f0fb';
}

.i-random:before {
  content: '\f0fc';
}

.i-recycle:before {
  content: '\f0fd';
}

.i-refresh:before {
  content: '\f0fe';
}

.i-registered:before {
  content: '\f0ff';
}

.i-repeat:before {
  content: '\f100';
}

.i-rotate-right:before {
  content: '\f100';
}

.i-retweet:before {
  content: '\f101';
}

.i-rocket:before {
  content: '\f102';
}

.i-rotate-left:before {
  content: '\f103';
}

.i-undo:before {
  content: '\f103';
}

.i-rss-square:before {
  content: '\f104';
}

.i-search:before {
  content: '\f105';
}

.i-search-minus:before {
  content: '\f106';
}

.i-search-plus:before {
  content: '\f107';
}

.i-share-alt:before {
  content: '\f108';
}

.i-share-square:before {
  content: '\f109';
}

.i-share-square-o:before {
  content: '\f10a';
}

.i-shield:before {
  content: '\f10b';
}

.i-shopping-bag:before {
  content: '\f10c';
}

.i-shopping-basket:before {
  content: '\f10d';
}

.i-shopping-cart:before {
  content: '\f10e';
}

.i-sign-in:before {
  content: '\f10f';
}

.i-sign-out:before {
  content: '\f110';
}

.i-signal:before {
  content: '\f111';
}

.i-sitemap:before {
  content: '\f112';
}

.i-slack:before {
  content: '\f113';
}

.i-sliders:before {
  content: '\f114';
}

.i-smile-o:before {
  content: '\f115';
}

.i-snapchat-ghost:before {
  content: '\f116';
}

.i-songs:before {
  content: '\f117';
}

.i-sort:before {
  content: '\f118';
}

.i-unsorted:before {
  content: '\f118';
}

.i-sort-amount-asc:before {
  content: '\f119';
}

.i-sort-amount-desc:before {
  content: '\f11a';
}

.i-sort-asc:before {
  content: '\f11b';
}

.i-sort-up:before {
  content: '\f11b';
}

.i-sort-desc:before {
  content: '\f11c';
}

.i-sort-down:before {
  content: '\f11c';
}

.i-soundcloud:before {
  content: '\f11d';
}

.i-spinner:before {
  content: '\f11e';
}

.i-spotify:before {
  content: '\f11f';
}

.i-square:before {
  content: '\f120';
}

.i-square-o:before {
  content: '\f121';
}

.i-star:before {
  content: '\f122';
}

.i-star-half:before {
  content: '\f123';
}

.i-star-half-empty:before {
  content: '\f124';
}

.i-star-half-full:before {
  content: '\f124';
}

.i-star-half-o:before {
  content: '\f124';
}

.i-star-o:before {
  content: '\f125';
}

.i-step-backward:before {
  content: '\f126';
}

.i-step-forward:before {
  content: '\f127';
}

.i-sticky-note:before {
  content: '\f128';
}

.i-sticky-note-o:before {
  content: '\f129';
}

.i-stop:before {
  content: '\f12a';
}

.i-stop-circle:before {
  content: '\f12b';
}

.i-stop-circle-o:before {
  content: '\f12c';
}

.i-street-view:before {
  content: '\f12d';
}

.i-suitcase:before {
  content: '\f12e';
}

.i-sun-o:before {
  content: '\f12f';
}

.i-tablet:before {
  content: '\f130';
}

.i-tag:before {
  content: '\f131';
}

.i-tags:before {
  content: '\f132';
}

.i-television:before {
  content: '\f133';
}

.i-tv:before {
  content: '\f133';
}

.i-th:before {
  content: '\f134';
}

.i-th-large:before {
  content: '\f135';
}

.i-th-list:before {
  content: '\f136';
}

.i-thumb-tack:before {
  content: '\f137';
}

.i-thumbs-down:before {
  content: '\f138';
}

.i-thumbs-o-down:before {
  content: '\f139';
}

.i-thumbs-o-up:before {
  content: '\f13a';
}

.i-thumbs-up:before {
  content: '\f13b';
}

.i-ticket:before {
  content: '\f13c';
}

.i-times-circle:before {
  content: '\f13d';
}

.i-times-circle-o:before {
  content: '\f13e';
}

.i-toggle-off:before {
  content: '\f13f';
}

.i-toggle-on:before {
  content: '\f140';
}

.i-trash:before {
  content: '\f141';
}

.i-trash-o:before {
  content: '\f142';
}

.i-trophy:before {
  content: '\f143';
}

.i-truck:before {
  content: '\f144';
}

.i-tumblr:before {
  content: '\f145';
}

.i-twitter:before {
  content: '\f146';
}

.i-twitter-square:before {
  content: '\f147';
}

.i-umbrella:before {
  content: '\f148';
}

.i-universal-access:before {
  content: '\f149';
}

.i-unlock:before {
  content: '\f14a';
}

.i-unlock-alt:before {
  content: '\f14b';
}

.i-upload:before {
  content: '\f14c';
}

.i-usb:before {
  content: '\f14d';
}

.i-user:before {
  content: '\f14e';
}

.i-user-secret:before {
  content: '\f14f';
}

.i-video-camera:before {
  content: '\f150';
}

.i-vimeo:before {
  content: '\f151';
}

.i-vimeo-square:before {
  content: '\f152';
}

.i-volume-off:before {
  content: '\f153';
}

.i-volume-down:before {
  content: '\f154';
}

.i-volume-middle:before {
  content: '\f155';
}

.i-volume-up:before {
  content: '\f156';
}

.i-whatsapp:before {
  content: '\f157';
}

.i-wheelchair:before {
  content: '\f158';
}

.i-wifi:before {
  content: '\f159';
}

.i-wordpress:before {
  content: '\f15a';
}

.i-wpforms:before {
  content: '\f15b';
}

.i-wrench:before {
  content: '\f15c';
}

.i-youtube:before {
  content: '\f15d';
}

.i-youtube-play:before {
  content: '\f15e';
}

.i-youtube-square:before {
  content: '\f15f';
}
