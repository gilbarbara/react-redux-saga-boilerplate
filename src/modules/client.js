// @flow
/**
 * Client
 * @module Client
 */

export class ServerError extends Error {
  response: Object;

  constructor(message?: string): Error {
    super(message);

    Error.captureStackTrace(this, ServerError);

    this.name = 'ServerError';

    return this;
  }
}

export function parseError(error: string): string {
  return error || 'Something went wrong';
}

/**
 * Fetch data
 *
 * @param {string} url
 * @param {Object} options
 * @param {string} [options.method] - Request method ( GET, POST, PUT, ... ).
 * @param {string} [options.payload] - Request body.
 * @param {Object} [options.headers]
 *
 * @returns {Promise}
 */
export function request(url: string, options: Object = {}): Promise<*> {
  const config = {
    method: 'GET',
    ...options,
  };
  const errors = [];

  if (!url) {
    errors.push('url');
  }

  if (!config.payload && (config.method !== 'GET' && config.method !== 'DELETE')) {
    errors.push('payload');
  }

  if (errors.length) {
    throw new Error(`Error! You must pass \`${errors.join('`, `')}\``);
  }

  const headers = {
    Accept: 'application/json',
    'Content-Type': 'application/json',
    ...config.headers,
  };

  const params: Object = {
    headers,
    method: config.method,
  };

  if (params.method !== 'GET') {
    params.body = JSON.stringify(config.payload);
  }

  return fetch(url, params).then(async response => {
    const contentType = response.headers.get('content-type');

    if (response.status > 299) {
      const error: Object = new ServerError(response.statusText);
      error.status = response.status;

      if (contentType && contentType.includes('application/json')) {
        error.response = await response.json();
      } else {
        error.response = await response.text();
      }

      throw error;
    } else {
      if (contentType && contentType.includes('application/json')) {
        return response.json();
      }

      return response.text();
    }
  });
}
