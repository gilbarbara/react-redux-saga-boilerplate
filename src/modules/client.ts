/**
 * Client
 * @module Client
 */
import { PlainObject } from 'types';

interface RequestOptions {
  body?: any;
  headers?: Record<string, string>;
  method?: 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE';
}

interface RequestError extends Error {
  status: number;
  response: any;
}

export function parseError(error: string): string {
  return error || 'Something went wrong';
}

/**
 * Fetch data
 *
 * @param {string} url
 * @param {Object} options
 * @param {string} [options.method] - Request method ( GET, POST, PUT, ... ).
 * @param {string} [options.payload] - Request body.
 * @param {Object} [options.headers]
 *
 * @returns {Promise}
 */
export function request(url: string, options: RequestOptions = {}): Promise<any> {
  const { body, headers, method }: RequestOptions = {
    method: 'GET',
    ...options,
  };

  const errors: string[] = [];

  if (!url) {
    errors.push('url');
  }

  if (!body && !['GET', 'DELETE'].includes(method)) {
    errors.push('payload');
  }

  if (errors.length) {
    throw new Error(`Error! You must pass \`${errors.join('`, `')}\``);
  }

  const params: RequestOptions = {
    headers: {
      Accept: 'application/json',
      'Content-Type': 'application/json',
      ...headers,
    },
    method,
  };

  if (method !== 'GET') {
    params.body = JSON.stringify(body);
  }

  return fetch(url, params).then(async response => {
    const text = await response.text();
    let content: string | PlainObject;

    try {
      content = JSON.parse(text);
    } catch {
      content = text;
    }

    if (response.status > 299) {
      const error = new Error(response.statusText) as RequestError;
      error.status = response.status;
      error.response = content;

      throw error;
    } else {
      return content;
    }
  });
}
