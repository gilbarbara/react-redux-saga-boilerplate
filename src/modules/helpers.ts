import { Reducer } from 'redux';
import { format } from 'date-fns';
import produce from 'immer';
import is from 'is-lite';

import {
  ActionCreator,
  ActionsMapReducer,
  GenericFunction,
  LoggerOptions,
  PlainObject,
  SortFunction,
  StoreAction,
} from 'types';

/**
 * Create an action
 */
export function createAction<T extends GenericFunction>(
  type: string,
  payloadCreator: T,
): ActionCreator<Parameters<T>, ReturnType<T>> {
  if (!payloadCreator) {
    throw new TypeError('Expected a function');
  }

  return (...args: any[]) => ({
    type,
    payload: payloadCreator(...args),
  });
}

/**
 * Create a reducer
 */
export function createReducer<State>(
  actionsMap: ActionsMapReducer<State>,
  defaultState: State,
): Reducer<State, StoreAction> {
  return (state = defaultState, action: StoreAction) =>
    produce(state, (draft: State) => {
      const fn = actionsMap[action.type];

      if (fn) {
        return fn(draft, action);
      }

      return draft;
    });
}

/**
 * Async Sleep
 */
export function sleep(seconds = 1) {
  return new Promise(resolve => setTimeout(resolve, seconds * 1000));
}

/**
 * Get unix timestamp in seconds
 */
export function getUnixtime(): number {
  return Math.floor(Date.now() / 1000);
}

/**
 * Check if cache is valid
 */
export function hasValidCache(lastUpdated: number, max = 10): boolean {
  if (!navigator.onLine) {
    return true;
  }

  return lastUpdated + max * 60 > getUnixtime();
}

export function keyMirror<T extends PlainObject>(input: T): { [K in keyof T]: K } {
  if (!is.plainObject(input)) {
    throw new TypeError('Expected an object');
  }

  const output: any = {};

  for (const key in input) {
    /* istanbul ignore else */
    if (!Object.prototype.hasOwnProperty.call(output, key)) {
      output[key] = key;
    }
  }

  return output;
}

/**
 * Log grouped messages to the console
 */
export function logger(
  type: string,
  title: string,
  data: unknown,
  options: LoggerOptions = {},
): void {
  /* eslint-disable no-console */
  const { collapsed = true, hideTimestamp = false, skip = false, typeColor = 'gray' } = options;
  const groupMethod = collapsed ? console.groupCollapsed : console.group;

  /* istanbul ignore else */
  if (process.env.NODE_ENV === 'development') {
    const parts = [`%c ${type}`];
    const styles = [`color: ${typeColor}; font-weight: lighter;`, 'color: inherit;'];
    const time = format(new Date(), 'HH:mm:ss');

    if (!hideTimestamp) {
      styles.push('color: gray; font-weight: lighter;');
    }

    parts.push(`%c${title}`);

    if (!hideTimestamp) {
      parts.push(`%c@ ${time}`);
    }

    /* istanbul ignore else */
    if (!skip && !window.HIDE_LOGS) {
      groupMethod(parts.join(' '), ...styles);
      console.log(data);
      console.groupEnd();
    }
  }
  /* eslint-enable no-console */
}

/**
 * Sort an array with localeCompare
 */
export function sortByLocaleCompare(
  key?: string,
  options: Intl.CollatorOptions & { descending?: boolean } = {},
): SortFunction {
  const { descending, ...compareOptions } = options;

  if (key) {
    if (descending) {
      return <T extends PlainObject>(left: T, right: T) =>
        right[key].toLowerCase().localeCompare(left[key].toLowerCase(), undefined, compareOptions);
    }

    return <T extends PlainObject>(left: T, right: T) =>
      left[key].toLowerCase().localeCompare(right[key].toLowerCase(), undefined, compareOptions);
  }

  if (descending) {
    return <T extends string>(left: T, right: T) =>
      right.toLowerCase().localeCompare(left.toLowerCase(), undefined, compareOptions);
  }

  return <T extends string>(left: T, right: T) =>
    left.toLowerCase().localeCompare(right.toLowerCase(), undefined, compareOptions);
}
