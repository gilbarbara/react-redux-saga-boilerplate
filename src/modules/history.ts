import { createBrowserHistory } from 'history';

const history = createBrowserHistory();

history.location = {
  ...history.location,
  state: {},
};

/* istanbul ignore next */
history.listen(() => {
  history.location = {
    ...history.location,
    state: history.location.state || {},
  };
});

const { go, goBack, push, replace } = history;

export { go, goBack, push, replace };
export default history;
