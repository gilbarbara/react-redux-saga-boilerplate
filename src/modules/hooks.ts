import { shallowEqual, useSelector } from 'react-redux';

import { StoreState } from 'types';

export function useShallowEqualSelector<TReturn>(selector: (state: StoreState) => TReturn) {
  return useSelector(selector, shallowEqual);
}
