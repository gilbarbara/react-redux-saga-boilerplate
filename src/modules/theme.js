export const headerHeight = 70;

export const appColor = '#00b4d5';

export const easing = 'cubic-bezier(0.35, 0.01, 0.77, 0.34);';

export default {
  breakpoints: [400, 768, 1024, 1280, 1920],
  palette: {
    primary: appColor,
  },
  button: {
    borderRadius: {
      xs: 4,
      lg: 28,
      xl: 32,
    },
    padding: {
      lg: [12, 28],
      xl: [14, 32],
    },
  },
};
