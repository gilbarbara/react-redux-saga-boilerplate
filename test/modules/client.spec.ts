import { parseError, request } from 'modules/client';

const fetchOptions = {
  headers: {
    'content-type': 'application/json',
  },
};

describe('modules/client', () => {
  beforeEach(() => {
    fetchMock.resetMocks();
  });

  describe('request', () => {
    it('should fail without param', () => {
      // @ts-ignore
      expect(() => request()).toThrow('Error! You must pass `url`');
    });

    it('should fail to a POST without payload', () => {
      expect(() => request('http://example.com/token', { method: 'POST' })).toThrow(
        'Error! You must pass `payload`',
      );
    });

    it('should handle a GET request', () => {
      fetchMock.mockResponseOnce(JSON.stringify({ hello: 'world' }), fetchOptions);

      return request('http://example.com/token').then(data => {
        expect(data).toMatchSnapshot();
      });
    });

    it('should handle a POST request', () => {
      fetchMock.mockResponseOnce('{}', {
        ...fetchOptions,
        status: 201,
      });

      return request('http://example.com/token', { method: 'POST', body: { a: 1 } }).then(data => {
        expect(data).toMatchSnapshot();
      });
    });

    it('should handle an error', () => {
      fetchMock.mockResponse(JSON.stringify({ message: 'Something went wrong' }), {
        status: 400,
      });

      return request('http://example.com/token').catch(error => {
        expect(error.toString()).toBe('Error: Bad Request');
        expect(error.status).toBe(400);
      });
    });

    it('should handle a server error with JSON response', () => {
      fetchMock.mockResponseOnce(JSON.stringify({ error: 'FAILED' }), fetchOptions);

      return request('http://example.com/token').catch(error => {
        expect(error.response).toMatchSnapshot();
      });
    });

    it('should handle a fetch error', () => {
      fetchMock.mockRejectOnce(new Error('Failed to Fetch'));

      return request('http://example.com/token').catch(error => {
        expect(error).toMatchSnapshot();
      });
    });
  });

  describe('parseError', () => {
    it('should handle messages', () => {
      expect(parseError('Failed')).toBe('Failed');
      // @ts-ignore
      expect(parseError()).toBe('Something went wrong');
    });
  });
});
