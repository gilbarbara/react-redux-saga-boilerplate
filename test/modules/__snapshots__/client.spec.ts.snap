// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/client request should handle a GET request 1`] = `
Object {
  "hello": "world",
}
`;

exports[`modules/client request should handle a POST request 1`] = `Object {}`;

exports[`modules/client request should handle a fetch error 1`] = `[Error: Failed to Fetch]`;
